### PR TITLE
Add use case for lowest memory consumption [8912]

### PR DESCRIPTION
**Pull request by imontesino**
Added a new section to use_cases and two code snippets to XMLTester.xml and DDSCodeTester.cpp

### CODE REVIEW

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/554/)

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/555/)

File: docs/fastdds/use_cases/use_cases.rst [hunk 1]
```diff
@@ -60,6 +60,10 @@ with distributed systems:
   this tutorial is an explanation of how to take full advantage of *Fast DDS* wide set of capabilities in a ROS 2
   project.
 
++ :ref:`use-case-reduceMemory`.
+  For use cases were available memory is scarce this example presents configuration options to reduce memory footprint
+  to a minimum by tweaking the :ref:`dds_layer_core_policy`.
```
Comment (EduPonz) [inline docs/fastdds/use_cases/use_cases.rst:65]: Qos Policies should be a link
```suggestion
  For use cases where available memory is scarce, *Fast DDS*  can configured to reduce memory footprint
  to a minimum by adjusting different  QoS policies.
```

File: docs/fastdds/use_cases/reduce_memory/reduce_memory.rst [hunk 2]
```diff
@@ -0,0 +1,97 @@
+.. include:: ../../../03-exports/aliases.include
+.. include:: ../../../03-exports/aliases-api.include
+
+.. _use-case-reduceMemory:
+
+Reduce the usage of memory
```
Comment (EduPonz) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:6]:
```suggestion
Reduce memory usage
```

File: docs/fastdds/use_cases/reduce_memory/reduce_memory.rst [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+==========================
```
Comment (EduPonz) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:7]: Adjust length to new title

File: docs/fastdds/use_cases/reduce_memory/reduce_memory.rst [hunk 4]
```diff
@@ -0,0 +1,97 @@
+.. include:: ../../../03-exports/aliases.include
+.. include:: ../../../03-exports/aliases-api.include
+
+.. _use-case-reduceMemory:
```
Comment (EduPonz) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:4]: Propagate
```suggestion
.. _use-case-reduce-memory:
```

File: docs/fastdds/use_cases/reduce_memory/reduce_memory.rst [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+Reduce the usage of memory
+==========================
+
+A great number of modern systems have tight constraints over available memory. Reducing the memory usage is critical for
```
Comment (EduPonz) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:9]: In general, please beak lines after periods. That way, future reviews will imply less modified lines, thus easing the reviews. RST build paragraphs it the sentences are only one line apart.

Comment (EduPonz) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:9]:
```suggestion
A great number of modern systems have tight constraints over available memory, making the reduction of memory usage to a minimum critical.
```

File: docs/fastdds/use_cases/reduce_memory/reduce_memory.rst [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+these applications. Reducing memory consumption of a Fast-DDS application can be achieved through a different number of
```
Comment (EduPonz) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:10]: Product name in italics + always *Fast DDS*
```suggestion
these applications. Reducing memory consumption of a *Fast DDS* application can be achieved through a different number of
```

File: docs/fastdds/use_cases/reduce_memory/reduce_memory.rst [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+approaches, mainly through architectural restructuring, but also via the limiting of resources and by avoiding static
```
Comment (EduPonz) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:11]:
```suggestion
approaches, mainly through architectural restructuring of the application, but also via limiting the resources the middleware utilizes, and by avoiding static
```

File: docs/fastdds/use_cases/reduce_memory/reduce_memory.rst [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+allocations.
+
+If a system has a low memory limit it is recommended to limit the maximum resources of the application to ensure that it
+will run properly. To lower the usage during runtime, allocations can be set dynamically to ensure the lowest amount of
+memory is always in use.
```
Comment (EduPonz) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:16]: I think this is the same info as the previous paragraph with a different formulation

File: docs/fastdds/use_cases/reduce_memory/reduce_memory.rst [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+QoS Adjustments
+---------------
+
```
Comment (EduPonz) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:20]: Avoid two title with no text. https://github.com/eProsima/all-docs/blob/master/CONTRIBUTING.md#text-structure point 7

File: docs/fastdds/use_cases/reduce_memory/reduce_memory.rst [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+ResourceLimitsQosPolicy
+^^^^^^^^^^^^^^^^^^^^^^^
+Limit the resources to the application's needs. The lowest it can be reduced to is one sample per |DataWriter-api| and
```
Comment (EduPonz) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:23]: I think we should explain here what this QoS does, and link to its sections in the docs. I think it's important to state what parameters are to be modified and why

File: docs/fastdds/use_cases/reduce_memory/reduce_memory.rst [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+|DataReader-api|.
+
+
+.. warning::
+
+   The value of |ResourceLimitsQosPolicy::max_samples-api| must be higher or equal to the value of
+   |ResourceLimitsQosPolicy::max_samples_per_instance-api|.
+
+   The value established for the :ref:`historyqospolicy` |HistoryQosPolicy::depth-api| must be lower or equal to the
+   value stated for |ResourceLimitsQosPolicy::max_samples_per_instance-api|.
```
Comment (EduPonz) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:33]: I'd put this as 2 bullet points, else the rendering is kinda ugly

File: code/DDSCodeTester.cpp [hunk 12]
```diff
@@ -3830,4 +3830,32 @@ void dds_usecase_examples()
         //!--
     }
 
+    {
+        //CONF-MEMORY-QOS-PUBSUB
+
+        ResourceLimitsQosPolicy resource_limits;
+        //The ResourceLimitsQosPolicy is default constructed with max_samples = 5000
```
Comment (EduPonz) [inline code/DDSCodeTester.cpp:3837]: This kind of comment are really helpful!

File: code/DDSCodeTester.cpp [hunk 13]
```diff
@@ -3830,4 +3830,32 @@ void dds_usecase_examples()
         //!--
     }
 
+    {
+        //CONF-MEMORY-QOS-PUBSUB
+
```
Comment (EduPonz) [inline code/DDSCodeTester.cpp:3835]:
```suggestion
```

File: code/DDSCodeTester.cpp [hunk 14]
```diff
@@ -3830,4 +3830,32 @@ void dds_usecase_examples()
...
+        //The ResourceLimitsQosPolicy is default constructed with max_instances = 10
+        //Change max_instances to the minimum
+        resource_limits.max_instances = 1;
+        //The ResourceLimitsQosPolicy is default constructed with max_samples_per_instance = 400
+        //Change max_samples_per_instance to the minimum
+        resource_limits.max_samples_per_instance = 1;
+        //The ResourceLimitsQosPolicy is default constructed with allocated_samples = 100
+        //No allocated samples
+        resource_limits.allocated_samples = 0;
+
```
Comment (EduPonz) [inline code/DDSCodeTester.cpp:3849]:
```suggestion
```

File: code/DDSCodeTester.cpp (hunk 12 above)
Comment (EduPonz) [inline code/DDSCodeTester.cpp:3837]: In general
```suggestion
        // The ResourceLimitsQosPolicy is default constructed with max_samples = 5000
```

File: docs/fastdds/use_cases/reduce_memory/reduce_memory.rst [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+|    :start-after: <!-->CONF-MEMORY-QOS-PUBSUB           |
+|    :end-before: <!--><-->                              |
+|    :lines: 2-3,5-                                      |
+|    :append: </profiles>                                |
++--------------------------------------------------------+
+
+Set Dynamic Allocation
+^^^^^^^^^^^^^^^^^^^^^^
+
+Using the dynamic settings of the :ref:`rtpsendpointqos` will prevent unecessery allocations. Lowest footprint is
```
Comment (EduPonz) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:57]:
```suggestion
Using the dynamic settings of the :ref:`rtpsendpointqos` will prevent unnecessary allocations. Lowest footprint is
```

File: docs/fastdds/use_cases/reduce_memory/reduce_memory.rst [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+| .. literalinclude:: /../code/XMLTester.xml             |
+|    :language: xml                                      |
+|    :start-after: <!-->CONF-MEMORY-QOS-ENDPOINTS        |
+|    :end-before: <!--><-->                              |
+|    :lines: 2-3,5-                                      |
+|    :append: </profiles>                                |
++--------------------------------------------------------+
+
+
+.. note::
```
Comment (EduPonz) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:83]: I think we can delete this note in favor of #178

File: docs/fastdds/use_cases/reduce_memory/reduce_memory.rst [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+|    :language: xml                                      |
+|    :start-after: <!-->CONF-MEMORY-QOS-PUBSUB           |
+|    :end-before: <!--><-->                              |
+|    :lines: 2-3,5-                                      |
+|    :append: </profiles>                                |
++--------------------------------------------------------+
+
+Set Dynamic Allocation
+^^^^^^^^^^^^^^^^^^^^^^
+
```
Comment (EduPonz) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:56]: I think it's worth mentioning what's the default Fast DDS behaviour and why it is like that. That way, the reader can better understand what does changing the memory policy entails.

File: docs/fastdds/use_cases/reduce_memory/reduce_memory.rst [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+Using the dynamic settings of the :ref:`rtpsendpointqos` will prevent unecessery allocations. Lowest footprint is
+achieved with |DYNAMIC_RESERVE_MEMORY_MODE-api| at the cost of higher allocation counts, for higher determinism at a
+small memory cost the |DYNAMIC_REUSABLE_MEMORY_MODE-api| option is available.
```
Comment (EduPonz) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:59]: I think we could clarify the difference between the two. Keep in mind that normal use-case readers probably have not read any section of the docs and came directly here to use the snippets they need. The use-case should explain what they need and also link to the appropriate sections

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/564/)

File: docs/fastdds/use_cases/use_cases.rst [hunk 19]
```diff
@@ -60,6 +60,10 @@ with distributed systems:
   this tutorial is an explanation of how to take full advantage of *Fast DDS* wide set of capabilities in a ROS 2
   project.
 
++ :ref:`use-case-reduceMemory`.
```
Comment (rsanchez15) [inline docs/fastdds/use_cases/use_cases.rst:63]: Update the link. The documentation doesn't pass the test right now.

File: docs/fastdds/use_cases/use_cases.rst [hunk 20]
```diff
@@ -60,6 +60,10 @@ with distributed systems:
...
+  For use cases where available memory is scarce, *Fast DDS*  can configured to reduce memory footprint
```
Comment (rsanchez15) [inline docs/fastdds/use_cases/use_cases.rst:64]:
```suggestion
  For use cases where available memory is scarce, *Fast DDS*  can be configured to reduce memory footprint
```

Comment (rsanchez15) [inline docs/fastdds/use_cases/use_cases.rst:64]: Memory doesn't have to be scarce, you simply have memory use constraints.

```suggestion
  For use cases with memory consumption constraints, *Fast DDS*  can configured to reduce memory footprint
```

File: docs/fastdds/use_cases/use_cases.rst [hunk 21]
```diff
@@ -60,6 +60,10 @@ with distributed systems:
...
+  to a minimum by adjusting different  QoS policies.
```
Comment (rsanchez15) [inline docs/fastdds/use_cases/use_cases.rst:65]:
```suggestion
  to a minimum by adjusting different QoS policies.
```

File: docs/fastdds/use_cases/reduce_memory/reduce_memory.rst [hunk 22]
```diff
@@ -0,0 +1,96 @@
+.. include:: ../../../03-exports/aliases.include
+.. include:: ../../../03-exports/aliases-api.include
+
+.. _use-case-reduce-memory:
+
+Reduce memory usage
+==========================
```
Comment (rsanchez15) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:7]:
```suggestion
Reduce memory usage
===================
```

File: docs/fastdds/use_cases/reduce_memory/reduce_memory.rst [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
++--------------------------------------------------------+
+
+
+.. note::
+    Memory allocation modes have different names for XML config files and C++ source code.
+
+    +--------------------------------------+----------------------------+
+    |C++                                   | XML                        |
+    +======================================+============================+
+    |PREALLOCATED_MEMORY_MODE              | PREALLOCATED               |
```
Comment (rsanchez15) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:89]: Use links to the API Reference, e.g. |PREALLOCATED_MEMORY_MODE-api|

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/567/)

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/568/)

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/569/)

File: docs/fastdds/use_cases/use_cases.rst (hunk 19 above)
Comment (rsanchez15) [reply docs/fastdds/use_cases/use_cases.rst]: This is not resolved.

File: docs/fastdds/use_cases/use_cases.rst (hunk 20 above)
Comment (rsanchez15) [reply docs/fastdds/use_cases/use_cases.rst]: This is not resolve either.

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/572/)

File: code/DDSCodeTester.cpp (hunk 12 above)
Comment (rsanchez15) [reply code/DDSCodeTester.cpp]: Add and space between `//` and the first letter of the comment. When it's said in general, it means that the comment applies to the entire PR. Please check that this format applies to all comments.

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/573/)

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/574/)

File: code/DDSCodeTester.cpp [hunk 24]
```diff
@@ -3830,4 +3830,35 @@ void dds_usecase_examples()
...
+        resource_limits.max_instances = 1;
+
+        // The ResourceLimitsQosPolicy is default constructed with max_samples_per_instance = 400
+        // Change max_samples_per_instance to the minimum
+        resource_limits.max_samples_per_instance = 1;
+        
+        // The ResourceLimitsQosPolicy is default constructed with allocated_samples = 100
+        // No allocated samples
+        resource_limits.allocated_samples = 0;
+
```
Comment (rsanchez15) [inline code/DDSCodeTester.cpp]:
```suggestion
```

File: docs/fastdds/use_cases/reduce_memory/reduce_memory.rst [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+minimum critical. Reducing memory consumption of a *Fast DDS* application can be achieved through a different number of
+approaches, mainly through architectural restructuring of the application, but also via limiting the resources the
+middleware utilizes, and by avoiding static allocations.
+
+.. |max_samples| replace:: |ResourceLimitsQosPolicy::max_samples-api|
+.. |max_instances| replace:: |ResourceLimitsQosPolicy::max_instances-api|
+.. |max_samples_per_instance| replace:: |ResourceLimitsQosPolicy::max_samples_per_instance-api|
+.. |allocated_samples| replace:: |ResourceLimitsQosPolicy::allocated_samples-api|
+.. |depth| replace:: |HistoryQosPolicy::depth-api|
+
```
Comment (rsanchez15) [inline docs/fastdds/use_cases/reduce_memory/reduce_memory.rst:19]: As @EduPonz suggests, QoS should be briefly explained here.
Check this again: https://github.com/eProsima/Fast-RTPS-docs/pull/177#discussion_r459352698

**Comment by richiware** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/575/)

**Pull request merged**